src/main/scala/mesosphere/marathon/core/election/CuratorElectionStream.scala

```diff
@@ -0,0 +1,243 @@
+    private lazy val latch = new LeaderLatch(client, curatorLeaderLatchPath, hostPort)
+    private var isStarted = false
+    @volatile private var _isCancelled = false
+
+    /* Long-poll trampoline-style recursive method which calls emitLeader() each time it detects that the leadership
+     * state has changed.
+     *
+     * Given instance A, B, C, Curator's Leader latch recipe only provides A the ability to be notified if it gains or
+     * loses leadership, but not if leadership transitions between B and C. This method allows us to monitor any change
+     * in leadership state.
+     *
+     * It is important that we re-register our watch _BEFORE_ we get the current leader. In Zookeeper,
+     * watches are one-time use only.
+     *
+     * The timeline of events looks like this
+     *
+     * 1) We register a watch
+     * 2) We query the current leader
+     * 3) We receive an event (indicating child removal / addition)
+     * 4) Repeat step 1
+     *
+     * By re-registering the watch before querying the state, we will not miss out on the latest leader change.
+     *
+     * We also have a retry and (very simple) back-off mechanism. This is because Curator's leader latch creates the
+     * initial leader node asynchronously. If we poll for leader information before this background hook completes, then
+     * a KeeperException.NoNodeException is thrown (which we handle, and retry)
+     */
+    def longPollLeaderChange(retries: Int = 0): Unit = singleThreadEC.execute { () =>
+      try {
+        if (latch.getState == LeaderLatch.State.STARTED)
+          client.getChildren
+            .usingWatcher(new CuratorWatcher {
+              override def process(event: WatchedEvent): Unit =
+                if (!_isCancelled) longPollLeaderChange()
+            })
+            .forPath(curatorLeaderLatchPath)
+        emitLeader()
+      } catch {
+        case ex: KeeperException.NoNodeException if retries < 100 =>
+          // Wait for node to be created
+          logger.info("retrying")
+          Thread.sleep(retries * 10L)
+          longPollLeaderChange(retries + 1)
+        case ex: Throwable =>
+          sq.fail(ex)
+      }
+    }
+
+    /**
+      * Emit current leader. Does not fail on connection error, but throws if multiple election candidates have the same
+      * ID.
+      */
+    private def emitLeader(): Unit = {
+      val currentLeader = leaderHostPortMetric.blocking {
+        val participants = try {
+          if (client.getState == CuratorFrameworkState.STOPPED)
+            Nil
+          else
+            latch.getParticipants.asScala.toList
+        } catch {
+          case ex: Throwable =>
+            logger.error("Error while getting current leader", ex)
+            Nil
+        }
+        if (participants.iterator.filter(_.getId == hostPort).size > 1)
+          throw new IllegalStateException(s"Multiple election participants have the same ID: ${hostPort}. This is not allowed.")
```

Comment:
I guess this should normally never happen?

Reply:
If you start two marathon instances with the same hostname, and the same port, then it would happen.

This would also be illegal.

Reply:
Curator's leader latch actually allows for this to happen. The trouble is, when querying the participant list, you can't tell which is your participant record, and which is some other participant record.

Just forbidding this state altogether kept it simple and allowed us to use a query mechanism to determine if we're the leader, or somebody else is.